src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.RegularCompilationTracker.cs

```diff
@@ -1096,40 +1096,50 @@ public Task<Checksum> GetDependentChecksumAsync(
-                            arg.self.ComputeDependentChecksumAsync(arg.SolutionState, c),
-                            arg: (self: this, compilationState.SolutionState)),
+                            arg.self.ComputeDependentChecksumAsync(arg.compilationState, c),
+                            arg: (self: this, compilationState)),
                         null);
                 }
 
                 return _lazyDependentChecksum.GetValueAsync(cancellationToken);
             }
 
-            private async Task<Checksum> ComputeDependentChecksumAsync(SolutionState solution, CancellationToken cancellationToken)
+            private async Task<Checksum> ComputeDependentChecksumAsync(
+                SolutionCompilationState solution, CancellationToken cancellationToken)
             {
                 using var _ = ArrayBuilder<Checksum>.GetInstance(out var tempChecksumArray);
 
+                // Mix in the SG information for this project.  That way if it changes, we will have a different
+                // checksum (since semantics could have changed because of this).
+                if (solution.SourceGeneratorExecutionVersionMap.Map.TryGetValue(this.ProjectState.Id, out var executionVersion))
+                    tempChecksumArray.Add(executionVersion.Checksum);
+
                 // Get the checksum for the project itself.
                 var projectChecksum = await this.ProjectState.GetChecksumAsync(cancellationToken).ConfigureAwait(false);
                 tempChecksumArray.Add(projectChecksum);
 
-                // Calculate a checksum this project and for each dependent project that could affect semantics for
-                // this project. Ensure that the checksum calculation orders the projects consistently so that we get
-                // the same checksum across sessions of VS.  Note: we use the project filepath+name as a unique way
-                // to reference a project.  This matches the logic in our persistence-service implemention as to how
-                // information is associated with a project.
-                var transitiveDependencies = solution.GetProjectDependencyGraph().GetProjectsThatThisProjectTransitivelyDependsOn(this.ProjectState.Id);
+                // Calculate a checksum this project and for each dependent project that could affect semantics for this
+                // project. We order the projects so that we are resilient to the underlying in-memory graph structure
+                // changing this arbitrarily.  We do not want that to cause us to change our semantic version.. Note: we
+                // use the project filepath+name as a unique way to reference a project.  This matches the logic in our
+                // persistence-service implementation as to how information is associated with a project.
+                var transitiveDependencies = solution.SolutionState.GetProjectDependencyGraph().GetProjectsThatThisProjectTransitivelyDependsOn(this.ProjectState.Id);
```

Comment:
Asking a general question which doesn't need to impact this PR:

Is there a reason this gets the full transitive list versus just asking the immediate project references and mixing in their checksums, which would themselves depend on their immediate project references...etc? Are we going to be generating extra dependency graph entries here even though we don't need to?

Reply:
i have a followup pr that does that :)